### PR TITLE
[WIP] Port CLI to Click

### PR DESCRIPTION
**Pull request by untitaker**
cont. #1446

### CODE REVIEW

**Comment by untitaker** [comment]
cc @sampsyo @duailibe: Adrian mentioned quickly that he wanted to implement a global click context. I'm working on integrating such functionality in Click itself. Since I still don't have full overview over beet's whole CLI code, I'd like to get some feedback on it, what you would like to see for your usecase: https://github.com/mitsuhiko/click/pull/354

**Comment by sampsyo** [comment]
Thanks for coalescing these changes, @untitaker! I will do a full review of the whole PR soon, but this looks good on a quick skim.

I'm also very curious about that Click PR, which I'll look closely at. On the beets side, we may consider doing something similar even if we don't use the Click-specific context machinery.

**Comment by untitaker** [comment]
My idea was rather to get this feature landed in Click before beets has to implement its own machinery.

**Comment by untitaker** [comment]
I just implemented command aliases in a way accessible for plugins -- please review.

**Comment by untitaker** [comment]
Click 5.0 will include a way to get the current context. However, it has to be manually pushed when spawning a new thread.

File: beets/ui/__init__.py [hunk 1]
```diff
@@ -1093,43 +1103,158 @@ def _open_library(config):
     return lib
 
 
-def _raw_main(args, lib=None):
+class BeetsCLI(click.MultiCommand):
+
+    def _commands(self, ctx):
+        # FIXME: Awful performance, cache per ctx
```
Comment (pradyunsg) [inline beets/ui/__init__.py]: Is this why a global context is wanted?

(PS: I didn't checkout and read, just browsing via Github...)

Comment (untitaker) [inline beets/ui/__init__.py]: Not really, we already get the ctx object passed.

Perhaps the cache should be stored on the ctx object itself though.

File: beets/ui/__init__.py [hunk 2]
```diff
@@ -773,6 +774,26 @@ def add_all_common_options(self):
 # succinct example program.
 
 
+# FIXME: kill original Subcommand in favor of this
```
Comment (untitaker) [inline beets/ui/__init__.py]: I feel like I need some help with this: Is backwards-compatibility for the plugin API important?

Comment (sampsyo) [inline beets/ui/__init__.py]: Good question. It would be nice if it's easy—and yes, I would love to look into writing that shim for compatibility—but not essential (as long as we port all the included plugins, of course).

File: .travis.yml [hunk 3]
```diff
@@ -1,9 +1,5 @@
 language: python
 
-branches:
-    only:
-        - master
-
```
Comment (jrobeson) [inline .travis.yml]: was this preemptively set in the first place? or was it set because of an actual problem? If the first, then maybe this should be its own PR.

Comment (jrobeson) [inline .travis.yml]: ping?

Comment (untitaker) [inline .travis.yml]: I suppose it was set in an effort to avoid multiple travis jobs (which happens when sending PRs from the source repo).

**Comment by sampsyo** [comment]
I've revived this branch and added an initial mechanism to automatically translate `optparse`-based code to use the new Click infrastructure. This should let us do this change incrementally and support third-party plugins.

My plan, after getting the translation working with better fidelity, is to land a version of this branch that relies on the translation as much as possible. That is, the first version won't use Click directly at all, where possible, so we'll have minimal changes to the commands themselves (both built-in and in plugins). Then, after the merge, we can work on migrating things over to using Click directly.

Here's what's outstanding:

* [ ] The shortcuts for adding `-p`, `-a`, and `-f` need to be overhauled for compatibility. Ideally, these would invoke the Click equivalents instead of the old optparse stuff.
* [ ] The duplicates plugin uses a "varargs callback" hack that has been removed. We need to figure out a shim or modify it to use Click directly.
* [ ] The `hide` option on `LegacySubcommand` is not yet respected.

Cleanup tasks:

* [ ] The main entry point needs to move from `_raw_main` to `main`.
* [ ] `AliasedSubcommand` needs some cleanup in the help department.
* [ ] `BeetsCommand` is a bit disorganized and underdocumented, and the setup code needs to be run only once.
* [ ] The long list of helpers for adding common options needs more documentation.
* [ ] We should consider moving all of this stuff to a new module, like `beets.ui.cli` or `beets.ui.parse` or something.
* [ ] We should check whether there's a better callback-based way to accomplish the `beet edit` shortcut than the current hack.

**Comment by sampsyo** [comment]
Solved the `varargs_callback` thing by not solving it: just switched to a more natural CLI option style.

**Comment by pradyunsg** [comment]
@sampsyo Just dropped in to say that it's great to see that you're working on this. :)

**Comment by sampsyo** [comment]
:smiley: Thanks, @pradyunsg!

I removed all the old bash completion stuff in favor of Click's built-in support. We could explore dropping our zsh script too and recommending zsh's `bashcompinit` compatibility layer.

**Comment by sampsyo** [comment]
Today I tackled a little more of the complexity surrounding the initial bootstrap and command dispatch stuff. I'm still not thrilled about it, but I think it will get better if we ever move to a "beets context" design wherein it's easier to run setup stuff on demand and with more explicit data flow.

Still plenty of little tasks to go before this is ready to merge, but we're getting closer.

**Comment by pkess** [comment]
Hey guys,

i just fixed some tests that were referencing the "old" raw_main function

**Comment by sampsyo** [comment]
Awesome; thank you! Using the `run_command` indirection looks like a good solution.

**Comment by pkess** [comment]
It seems to me like this warning click prints:
```
Warning: Click detected the use of the unicode_literals __future__ import.  This is heavily discouraged because it can introduce subtle bugs in your code.  You should instead use explicit u"" literals for your unicode strings.  For more information see http://click.pocoo.org/python3/
```
breaks many tests.
should we follow the warning and remove import of unicode_literals ?

**Comment by sampsyo** [comment]
@pkess Yep! That's one of the next steps. A couple of notes:

- We should do it on the master branch. This will give that change some time to settle independently. It could break a lot of things, so better to debug it by itself than at the same time as the click transition.
- We need to be very careful about adjusting all the literals to use explicit u and b prefixes as we remove the flag. Most literals should probably be explicitly annotated, with only a minority using unprefixed "native" strings.

**Comment by pkess** [comment]
OK, so maybe I will open a PR tomorrow with fixed literals. We can merge it to both master and click then

**Comment by sampsyo** [comment]
Awesome! Thanks for looking into it!

**Comment by sampsyo** [comment]
Things are getting closer. Here's a rejuvenated to-do list:

- [ ] Turn off `unicode_literals` (on main branch).
- [ ] Fix tests.
- [ ] Hidden `beet completion` command for backwards compatibility. (The most recent Click has support for hidden commands, but we may want to duplicate this functionality for better compatibility.)
- [ ] The implementations are still rather messy of the decorators for common command-line options.
- [ ] We need to reconsider the `Context` class mechanism for passing information from the top-level invocation to individual commands. In particular, the name is confusing, especially since there's an identically-named `pass_context` as in click

**Comment by untitaker** [comment]
Rebased.

**Comment by sampsyo** [comment]
Thank you, and sorry for my whitespace shenanigans. (I can't remember now why that seemed important.) It's on my to-do list to finish this up.

**Comment by jrobeson** [comment]
I'd love to work on this, but it's difficult to rebase on master, and tons of the first set of code gets reverted by future changes.  Can I squash this down a bit?  Although, preferably it'd be nice if either @untitaker or @duailibe cleaned it up a bit.  Alternatively, maybe it should be reimplemented from scratch. 

On a related note, should we force click 5? or continue on with the default install of click 6?

**Comment by untitaker** [comment]
It probably makes sense to either squash this into one commit or rewrite it from scratch. I think it was problematic that master moved at all while this was ongoing.

**Comment by jrobeson** [comment]
@untitaker : so what about click 5 vs click 6?

**Comment by jrobeson** [comment]
@sampsyo : as we move towards py3, is the legacy compatibility something that we should be dealing with anymore? All external plugins are likely to break anyways.   Maybe we should just offer help in porting them instead.

**Comment by jrobeson** [comment]
@untitaker can you reply to my query about click 5 vs 6?

**Comment by untitaker** [comment]
Just use Click 6 for now although version 7 will likely break compatibility in
a minor way.

**Comment by jrobeson** [comment]
I've got this rebased and have started working on it. 

How do we want our stdin related tests to act?

Here's an example:

```python
with control_stdin('y'):
    self.run_convert() 
```

I have it currently working like:

```python
self.run_convert(input='y')
```

So what do we really want here? Do we wanna keep a context manager style? or rely on click to do handle all this? Any other suggestions?

**Comment by jrobeson** [comment]
Also decargs no longer deals with non-ascii args correctly. You can see the results (even without the rebase) in the current failed tests: https://travis-ci.org/beetbox/beets/jobs/133517884#L349

**Comment by sampsyo** [comment]
Good point. Perhaps decargs should be a no-op under click? We (hopefully) shouldn't need to decode arguments anymore.

**Comment by jrobeson** [comment]
the rebase was approved by @sampsyo in #2190 and now pushed here.

**Comment by jrobeson** [comment]
so, we probably don't want to change tests here too much do we? So what do you think about adding `input` as a keyword argument to `run_command`/`run_with_output`? or should we go another way?

**Comment by jrobeson** [comment]
aren't the interactive issues mostly with the importer? We mostly don't use those with `run_command`

**Comment by sampsyo** [comment]
Yep, good point.

**Comment by jrobeson** [comment]
@sampsyo:  I am going to clarify a bit.  ATM I was under the impression that we wanted the test suite to pass in both click branch and master with as little modification as possible.  I was hoping to keep test suite modifications to  totally different implementations like the help command, bash completion, and CommonOptionsParser.

We're actually doing rather well on this, and ATM i don't think we have to modify users of `control_stdin`/`capture_stdout` in places like the importer. However, when it comes to testing the actual commands via `run_with_output`/`run_command` then `capture_stdout`/`control_stdin` no longer work as expected.

**Comment by sampsyo** [comment]
Aha; thanks for clarifying! That goal does make sense, I suppose, since it should make the merge cleaner.

Sorry for being slow, but I don't quite see how `capture_stdout`/`control_stdin` will break in some cases but not others. Is it an encoding thing?

File: beets/ui/__init__.py [hunk 4]
```diff
@@ -1180,55 +863,108 @@ def _open_library(config):
...
+        if str_args[0] == 'config' and ('-e' in str_args or
+                                        '--edit' in str_args):
+            from beets.ui.commands import config_cmd
+            command = config_cmd._to_click()
+            return str_args[0], command, args[1:]
+
+        # In the normal case, configure beets and then load the set of
+        # subcommands immediately before continuing with the resolution
+        # logic.
+        config = _configure(ctx.params)
```
Comment (jrobeson) [inline beets/ui/__init__.py]: ths is is eating the verbose configuration from the logging tests (and I'm sure other places). 

Click sets verbose to 0 since we have no default and then writes right over the parameter.

```python
    def test_command_level1(self):
        self.config['verbose'] = 1
        with helper.capture_log() as logs:
            self.run_command('dummy')
```
Do we want to use dynamic defaults for these parameters directly in `click.option`? or solve it at a higher level? I still don't get click well enough to know what would be best.

Our current configuration mechanism is kind of spooky action at a distance.

**Comment by jrobeson** [comment]
no, it's not an encoding thing. It's just that the importer has it's own runner (TestImportSession or something)  that doesn't rely the click test runner. Only `run_command`/`run_with_output` do.

**Comment by jrobeson** [comment]
The other cases for `capture_stdout` are things like 

```python
with capture_stdout as output:
    ui.some_command()
```

Those don't rely on the click test runner either.

**Comment by sampsyo** [comment]
Ah, got it, thanks! So it's just when we invoke commands using their names, when it makes sense to switch to the Click entry point for tests. This makes sense now. This direction seems right to me.

File: beets/ui/__init__.py (hunk 4 above)
Comment (jrobeson) [inline beets/ui/__init__.py]: I must not be quite correct in my analysis, as adding `-vv` to `self.run_command('dummy')`  makes it weirder.   

Failure without `-v`:
```sh
tox -e py35-test test/test_logging.py:LoggingLevelTest.test_command_level1  --  -xsv

======================================================================
FAIL: test_command_level1 (test.test_logging.LoggingLevelTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/johnny/src/beets/test/test_logging.py", line 105, in test_command_level1
    self.assertIn(u'dummy: debug cmd', logs)
AssertionError: 'dummy: debug cmd' not found in ['dummy: info cmd', 'dummy: warning cmd']
```

Failure with `-v`
```sh
tox -e py35-test test/test_logging.py:LoggingLevelTest.test_command_level1  --  -xsv
======================================================================
FAIL: test_command_level1 (test.test_logging.LoggingLevelTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/johnny/src/beets/test/test_logging.py", line 103, in test_command_level1
    self.assertIn(u'dummy: warning cmd', logs)
AssertionError: 'dummy: warning cmd' not found in []

```

Comment (sampsyo) [inline beets/ui/__init__.py]: Whoa; that's weird. Adding -v hides the warning-level message? :scream:

**Comment by jtpavlock** [comment]
Is this still something we'd like to implement?

**Comment by untitaker** [comment]
I'm no longer driving this PR for sure.
